Support for CSS text-transform: full-size-kana

mdubet opened the pull request:
#### 8ca11a6d7868e1c06e58c38eaa8349d5d39e1345
<pre>
Support for CSS text-transform: full-size-kana
<a href="https://bugs.webkit.org/show_bug.cgi?id=251934">https://bugs.webkit.org/show_bug.cgi?id=251934</a>
rdar://105181094

Reviewed by Tim Nguyen and Darin Adler.

The mapping from small kanas to full size kanas
is specified at <a href="https://drafts.csswg.org/css-text-3/#small-kana">https://drafts.csswg.org/css-text-3/#small-kana</a>

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-text/parsing/text-transform-computed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-text/parsing/text-transform-valid-expected.txt:
* LayoutTests/inspector/css/getSupportedCSSProperties-expected.txt:
* Source/WebCore/css/CSSPrimitiveValueMappings.h:
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/CSSValueKeywords.in:
* Source/WebCore/rendering/RenderText.cpp:
(WebCore::convertToFullSizeKana):
(WebCore::applyTextTransform):
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleConstants.cpp:
(WebCore::operator&lt;&lt;):
* Source/WebCore/rendering/style/RenderStyleConstants.h:

Canonical link: <a href="https://commits.webkit.org/260307@main">https://commits.webkit.org/260307@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3ef89579dd975d547b725c3f8aeaf1b2b28fef2c

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/107895 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/16950 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/40778 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/117027 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/116377 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/111782 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/18339 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/8267 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/100078 "Built successfully") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/113657 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/13841 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/97061 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/41570 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/95749 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/28699 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/83400 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/9875 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/30047 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/10582 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/84/builds/6939 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/16027 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/49636 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7131 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/12154 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/rendering/RenderText.cpp:
```diff
@@ -1329,6 +1329,55 @@ UChar RenderText::previousCharacter() const
+        0xFF71, 0xFF72, 0xFF73, 0xFF74, 0xFF75, 0xFF94, 0xFF95, 0xFF96, 0xFF82
+    };
+
+    StringBuilder converted;
```
I know nothing about this code/feature but isn't there a need for a fast path to avoid re-creating a brand new String in the case where no kana is found?

Or is it just not worth optimizing for in this case?

mdubet replied:
This feature is really niche, I don't think we should look for any optimisation there.

nt1m commented on Source/WebCore/rendering/RenderText.cpp:
```diff
@@ -1329,6 +1329,55 @@ UChar RenderText::previousCharacter() const
+        else
+            converted.append(UChar { character });
+    }
+    return converted.toString();
```
Can we do the same optimization as `capitalize` to return the original string if unchanged?

nt1m commented on Source/WebCore/rendering/RenderText.cpp:
```diff
@@ -1329,6 +1329,55 @@ UChar RenderText::previousCharacter() const
+        // Linear search is fast enough for a ~50 elements list.
+        auto founded = std::find(std::begin(smallKanas), std::end(smallKanas), character);
+        if (founded != std::end(smallKanas))
+            converted.append(UChar { fullSizeKanas[std::distance(std::begin(smallKanas), founded)] });
```
nit: found is correct english.

nt1m reviewed (commented):
Also, this commit doesn't implement combining uppercase/lowercase/capitalize with full-size-kana as specified. Are you planning to fix this in a followup commit?

szewai commented on Source/WebCore/rendering/RenderText.cpp:
```diff
@@ -1329,6 +1329,55 @@ UChar RenderText::previousCharacter() const
+// https://www.w3.org/TR/css-text-3/#small-kana
+static String convertToFullSizeKana(const String& string)
+{
+    static const uint16_t smallKanas[] = {
```
Might use `static constexpr std::pair<uint16_t, uint16_t> mapping[]` and converted to `SortedArrayMap`.

mdubet replied:
This feature is so rare that we might prefer optimising compilation time instead of run time for this.

litherum replied:
I think I disagree. Some pages might have this feature enabled gazillions of times on gazillions of content, and we want such pages to work well too.

mdubet replied:
It's true, and they will work properly indeed, with or without a `constexpr`.

litherum replied:
work _well_ => with good performance

mdubet replied:
Changed to `constexpr SortedArrayMap`

nt1m reviewed (approved):
It would be nice if @darinadler can look at the `convertToFullSizeKana` string manipulation bits if he has time.

litherum commented on Source/WebCore/rendering/RenderText.cpp:
```diff
@@ -1329,6 +1329,55 @@ UChar RenderText::previousCharacter() const
+        // ぁ   ぃ      ぅ      ぇ      ぉ      っ      ゃ      ゅ      ょ
+        0x3041, 0x3043, 0x3045, 0x3047, 0x3049, 0x3063, 0x3083, 0x3085, 0x3087,
+        // ゎ   ゕ      ゖ
+        0x308E, 0x3095, 0x3096,
```
Can we try very hard not to put these kinds of tables in WebKit? We should see if there's something in ICU that we can use. If not, we should make sure there's a feature request open on upstream ICU for it, and discuss with Apple's ICU team to see if we can come up with an interim solution.

mdubet replied:
There is a csswg issue open about this: https://github.com/w3c/csswg-drafts/issues/8442

In the meantime, having it in WebKit seems a reasonable solution (note that Firefox is doing the same).

litherum replied:
Have we at least started a conversation with Apple's ICU maintainers?

darinadler replied:
ICU includes a transliterator called `Fullwidth-Halfwidth`. Is the problem with this performance, correctness, or did we just not try it?

mdubet replied:
We are actually gonna use the ICU `Fullwidth-Halfwidth` transliterator for `text-transform: full-width`.

My understanding	of font stuff is quite limited, but I don't think it's the same mapping from small to full-size Kana. And this specific mapping is actually only defined in the CSS Text spec (but again, I might be wrong and I would strongly prefer not having this kind of mapping table inside WebKit).

https://www.w3.org/TR/css-text-3/#text-transform-mapping

> The mappings for [small Kana](https://www.w3.org/TR/css-text-3/#kana-small) to [full-size Kana](https://www.w3.org/TR/css-text-3/#kana-full-size) are defined in [Appendix G: Small Kana Mappings](https://www.w3.org/TR/css-text-3/#small-kana).

darinadler replied:
I see. So surprised this is not in ICU! Thanks for pointing out: full-width to half-width is not the same as small Kana to full-size Kana, my mistake.

litherum commented on Source/WebCore/rendering/RenderText.cpp:
```diff
@@ -1329,6 +1329,55 @@ UChar RenderText::previousCharacter() const
+
+    StringBuilder result;
+    for (const auto& character : string.charactersWithoutNullTermination()) {
+        // Linear search is fast enough for a ~50 elements list.
```
Is linear search actually faster than binary search here? I would be surprised if this was actually true.

mdubet replied:
The comment says "fast enough", not "faster". 

But linear search is often actually faster than binary search (or hashing) for small size. I don't think it's worth investigating/profiling this now though, but we can leave a FIXME if we really want to go down this road at some point.

darinadler replied:
Yes, that’s why `SortedArrayMap` does linear search for smaller arrays and binary search for larger. We profiled different cutoffs when creating it.

darinadler replied:
This should use `SortedArrayMap` rather than rolling its own.

litherum reviewed (commented):
WebKit ideally shouldn't be in the business of dealing with raw characters; that's what lower-level frameworks like ICU or Core Foundation are for. I think we should try very hard not to add more character tables to WebKit.

darinadler commented on Source/WebCore/rendering/RenderText.cpp:
```diff
@@ -1329,6 +1329,55 @@ UChar RenderText::previousCharacter() const
+    };
+
+    StringBuilder result;
+    for (const auto& character : string.charactersWithoutNullTermination()) {
```
This is an unnecessarily slow way to iterate characters. We have more efficient ones that don’t involve memory allocation that are incredibly easy to use. `StringView` has one for iterating UTF-16 code units and another for iterating characters (code points).

mdubet replied:
Changed to use `StringView::codePoints()`

darinadler commented on Source/WebCore/rendering/RenderText.cpp:
```diff
@@ -1329,6 +1330,73 @@ UChar RenderText::previousCharacter() const
+    StringBuilder result;
+    auto codePoints = StringView { string }.codePoints();
+    for (const auto character : codePoints) {
+        if (auto found = sortedMap.tryGet(character))
```
This is going to chop off the high 16 bits of the `UChar32`, and that might do the wrong thing in theory if the characters’s low 16 bits happened to be the same as an entry in the table. Since all the characters we are searching for are in the `U+0000-U+FFFF` range, it would be better to use `codeUnits` instead of `codePoints`, which would sidestep that problem, and a bonus is that we would not have to convert all surrogate pairs into `UChar32` and back again.

mdubet replied:
Fixed

darinadler commented on Source/WebCore/rendering/RenderText.cpp:
```diff
@@ -1329,6 +1330,73 @@ UChar RenderText::previousCharacter() const
+static String convertToFullSizeKana(const String& string)
+{
+    // https://www.w3.org/TR/css-text-3/#small-kana
+    static constexpr std::pair<uint16_t, uint16_t> kanasMap[] = {
```
We could use `std::pair<UChar, UChar>` here and avoid having to type cast below.

mdubet replied:
Fixed

darinadler commented on Source/WebCore/rendering/RenderText.cpp:
```diff
@@ -1329,6 +1330,73 @@ UChar RenderText::previousCharacter() const
+        { 0xFF6E, 0xFF96 },
+        { 0xFF6F, 0xFF82 }
+    };
+    constexpr auto sortedMap = SortedArrayMap { kanasMap };
```
Not sure why the one above is `static constexpr` and this is just `constexpr`. I believe they should both be the same since there are no different considerations for the two. I also suggest omitting `auto` and `=` and just writing `static constexpr SortedArrayMap sortedMap { kanasMap };`.

mdubet replied:
Changed to use `static`.

mdubet committed:
Support for CSS text-transform: full-size-kana
https://bugs.webkit.org/show_bug.cgi?id=251934
rdar://105181094

Reviewed by Tim Nguyen and Darin Adler.

The mapping from small kanas to full size kanas
is specified at https://drafts.csswg.org/css-text-3/#small-kana

* LayoutTests/TestExpectations:
* LayoutTests/imported/w3c/web-platform-tests/css/css-text/parsing/text-transform-computed-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/css/css-text/parsing/text-transform-valid-expected.txt:
* LayoutTests/inspector/css/getSupportedCSSProperties-expected.txt:
* Source/WebCore/css/CSSPrimitiveValueMappings.h:
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/CSSValueKeywords.in:
* Source/WebCore/rendering/RenderText.cpp:
(WebCore::convertToFullSizeKana):
(WebCore::applyTextTransform):
* Source/WebCore/rendering/style/RenderStyle.h:
* Source/WebCore/rendering/style/RenderStyleConstants.cpp:
(WebCore::operator<<):
* Source/WebCore/rendering/style/RenderStyleConstants.h:

Canonical link: https://commits.webkit.org/260307@main
